h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -522,74 +521,132 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
+                        colIndex[rowPointer][colPointer++] = di.getCategoricalId(j, Float.NaN);
                     } else {
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
+                        colIndex[rowPointer][colPointer++] = di.getCategoricalId(j, vecs[j].at8(i));
                     }
                     nonZeroCount++;
                 }
 
                 for (int j = 0; j < di._nums; ++j) {
+                    if(colPointer == SPARSE_MATRIX_DIM){
+                        colPointer = 0;
+                        rowPointer++;
+                    }
                     float val = (float) vecs[di._cats + j].at(i);
                     if (val != 0) {
-                        data[currentRow][currentCol] = val;
-                        colIndex[currentRow][currentCol++] = di._catOffsets[di._catOffsets.length - 1] + j;
+                        data[rowPointer][colPointer] = val;
+                        colIndex[rowPointer][colPointer++] = di._catOffsets[di._catOffsets.length - 1] + j;
+                        if(!foundNonZero){
+                            foundNonZero = true;
+                            for (int k = lastNonZeroRow; k < actualRows; k++) {
+                                rowHeaders[rowHeaderRowPointer][rowHeaderColPointer++] = nonZeroCount;
+                            }
+                            lastNonZeroRow = actualRows;
+                        }
                         nonZeroCount++;
                     }
                 }
-
-                rowHeaders[0][++actualRows] = nonZeroCount;
-
+                
                 rwRow = setResponseAndWeight(w, resp, weights, respVec, rwRow, i);
             }
         }
-
         return actualRows;
     }
 
     private static int initializeFromChunks(Chunk[] chunks, int weight, DataInfo di, int actualRows, long[][] rowHeaders, float[][] data, int[][] colIndex, int respIdx, float[] resp, float[] weights) {
         int nonZeroCount = 0;
-        int currentRow = 0;
+        int rowPointer = 0;
         int currentCol = 0;
         int rwRow = 0;
 
+        int rowHeaderRowPointer = 0;
+        int rowHeaderColPointer = 0;
+        int lastNonZeroRow = 0;
+
         for (int i = 0; i < chunks[0].len(); i++) {
             if (weight != -1 && chunks[weight].atd(i) == 0) continue;
+            actualRows++;
+            if(rowHeaderColPointer == SPARSE_MATRIX_DIM){
+                rowHeaderColPointer = 0;
+                rowHeaderRowPointer++;
+            }
+            boolean foundNonZero = false;
 
             for (int j = 0; j < di._cats; ++j) {
-                data[currentRow][currentCol] = 1; //one-hot encoding
+                if(currentCol == SPARSE_MATRIX_DIM){
+                    currentCol = 0;
+                    rowPointer++;
+                }
+                data[rowPointer][currentCol] = 1; //one-hot encoding
+                if(!foundNonZero){
```

Comment:
The FIRST non-zero character belongs to these row-related indices (pointing to the index of the non-zero element just discovered in the array).